price-feeder/oracle/client/client.go

```
@@ -0,0 +1,274 @@
+)
+
+type (
+	CosmosChain struct {
+		ChainID             string
+		KeyringBackend      string
+		KeyringDir          string
+		KeyringPass         string
+		TMRPC               string
+		RPCTimeout          time.Duration
+		OracleAddr          sdk.AccAddress
+		OracleAddrString    string
+		ValidatorAddr       sdk.ValAddress
+		ValidatorAddrString string
+		Encoding            umeeparams.EncodingConfig
+		GasPrices           string
+		GasAdjustment       float64
+		GRPCEndpoint        string
+	}
+
+	passReader struct {
+		pass string
+		buf  *bytes.Buffer
+	}
+)
+
+func newPassReader(pass string) io.Reader {
+	return &passReader{
+		pass: pass,
+		buf:  new(bytes.Buffer),
+	}
+}
+
+func (r *passReader) Read(p []byte) (n int, err error) {
+	n, err = r.buf.Read(p)
+	if err == io.EOF || n == 0 {
+		r.buf.WriteString(r.pass + "\n")
+
+		n, err = r.buf.Read(p)
+	}
+
+	return n, err
+}
+
+type OracleClient struct {
+	CosmosChain       CosmosChain
+	KeyringPassphrase string
+	Ctx               client.Context
+	Factory           tx.Factory
+}
+
+// Pre-vote and vote are separated out mainly for readability
+// Prevote doesn't need the timeout functionality that vote needs,
+// Because of the block timing validation on the node side
+
+// Ref : https://github.com/terra-money/oracle-feeder/blob/baef2a4a02f57a2ffeaa207932b2e03d7fb0fb25/feeder/src/vote.ts#L230
+
+func (oc OracleClient) BroadcastPrevote(msgs ...sdk.Msg) error {
+	return tx.BroadcastTx(oc.Ctx, oc.Factory, msgs...)
+}
+
+// Need to try this continually, since err is either timing issue or whitelist error
+// Ref : https://github.com/terra-money/core/blob/746a15f1bd83d62cd284e4af9471dc58701b3e33/x/oracle/keeper/msg_server.go#L89
+
+func (oc OracleClient) BroadcastVote(nextBlockHeight int, timeoutHeight int, msgs ...sdk.Msg) error {
+
+	var err error
+	maxBlockHeight := nextBlockHeight + timeoutHeight
+	lastCheckHeight := nextBlockHeight - 1
+	height := 0
+
+	for height == 0 && lastCheckHeight < maxBlockHeight {
```

Can we add a comment on what this loop is doing and what the timeouts mean? I feel uneasy about seeing random sleeps.

---

Got these timeouts from terra's feeder ( https://github.com/terra-money/oracle-feeder/blob/baef2a4a02f57a2ffeaa207932b2e03d7fb0fb25/feeder/src/vote.ts#L230 ) - seems like it takes time to be in the correct voting period. I added a retry here too ( why the error is ignored in this func ), but I'll remove that until we have the oracle module wired up 👍